internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,349 @@
+		return errors.Trace(err)
+	}
+
+	defer func() {
+		if err := w.cleanupPublicKey(ephemeralPublicKey); err != nil {
+			w.logger.Errorf("Error cleaning up ephemeral public key: %v", err)
+		}
+	}()
+
+	if err := w.pipeConnectionToSSHD(ctrlAddress, reqParams.Password); err != nil {
+		w.logger.Errorf("Error piping connection: %v", err)
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// pipeConnectionToSSHD initiates the connection back to the controller and pipes
+// it over to the local SSHD. This call blocks until the connection has finished.
+func (w *sshSessionWorker) pipeConnectionToSSHD(ctrlAddress, password string) error {
+	conn, err := w.connectionGetter.GetSSHConnection(password, ctrlAddress)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer conn.Close()
+
+	sshdConn, err := w.connectionGetter.GetSSHDConnection()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer sshdConn.Close()
+
+	eg := errgroup.Group{}
+
+	eg.Go(func() error {
+		// sshd -> conn
+		_, err := io.Copy(conn, sshdConn)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	eg.Go(func() error {
+		// conn -> sshd
+		_, err = io.Copy(sshdConn, conn)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := eg.Wait(); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// cleanupPublicKey finds the fingerprint of the provided key and attempts to delete the key
+// from authorized_keys2.
+func (w *sshSessionWorker) cleanupPublicKey(ephemeralPublicKey string) error {
+	fingerprint, _, err := jujussh.KeyFingerprint(ephemeralPublicKey)
+	if err != nil {
+		w.logger.Errorf("failed to get fingerprint of ephemeral public key: %v", err)
+		return errors.Trace(err)
+	}
+
+	if err := jujussh.DeleteKeysFromFile(ControllerSSHUser, AuthorizedKeysFile, []string{fingerprint}); err != nil {
+		w.logger.Errorf("Error deleting ephemeral public key: %v", err)
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+type ConnectionGetter interface {
+	GetSSHConnection(password, ctrlAddress string) (net.Conn, error)
+	GetSSHDConnection() (net.Conn, error)
+}
+
+type connectionGetter struct {
+	logger Logger
+}
+
+// NewConnectionGetter returns a struct capable of initating SSH connections to two places.
+//
+//  1. The controller's SSH server
+//  2. The local SSHD on the machine
+//
+// The consumer is expect to pipe these connections together.
+func NewConnectionGetter(l Logger) *connectionGetter {
+	return &connectionGetter{l}
+}
+
+// GetSSHConnection initiates an SSH connection to the target ctrlAddress.
+func (w *connectionGetter) GetSSHConnection(password, ctrlAddress string) (net.Conn, error) {
+	// TODO(ale8k): Watch will return host key in subsequent PR.
+	sshConfig := &ssh.ClientConfig{
+		User:            ControllerSSHUser,
+		Auth:            []ssh.AuthMethod{ssh.Password(password)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // TODO(ale8k): Fill in host key here
```

Comment:
We can put this in the connection request if needed. Or possibly we can put it in agent.conf, just as the unit agents know the controller's custom CA cert, they should be able to get the controller public host key from the same place.

Reply:
I think in the request is better, I'll make a card later (alongside the truncating auth keys).

Reply:
@SimoneDutto Please verify if a card exists for this or create one if necessary. I think it makes the most sense to push the controller's public host key onto all machines once when we provision the machines.

Reply:
fine for a followup